Added SharpSite.Plugins project and refactored plugin system

default-writer opened the pull request:
I added several ideas of how i see plugin management system.

csharpfritz commented:
You've got some neat ideas here, and some things that I wanted to plan for, but I need a bit more of a description of what the contribution is since its not connected to an issue that describes the requirement.

default-writer commented:
> You've got some neat ideas here, and some things that I wanted to plan for, but I need a bit more of a description of what the contribution is since its not connected to an issue that describes the requirement.

Ok, I plan to separate plugin system, related to #106 #107 #118 and #166 (wip, packaging tool)

default-writer commented:
@csharpfritz ready to codereview

1) moved PluginManifest from Abstrations to Plugins 
2) added according to example PluginAssemblyLoadContext class
3) added PluginAssembly class
4) added PluginAssemblyManager class for managing plugin context (loading / unloading, replacing in-memory)
5) added PluginException for handling specific plugin exceptions (not a file io or memory io)
6) replaced PluginManifest.id with PluginManifest.Id and added deserialization attribute (C# style guileines)
7) added Plugin class for binary container of data with some naming

csharpfritz commented:
Thank you for the code, this is a big refactoring we were going to need to do
